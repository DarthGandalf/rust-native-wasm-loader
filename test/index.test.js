import webpack from 'webpack';
import MemoryFS from 'memory-fs';
import path from 'path';

describe('rust-native-wasm-loader', () => {
  it('loads a simple cargo project', async () => {
    jest.setTimeout(100000);

    const options = {
      release: true,
    };

    const preRules = [{
      loader: 'wasm-loader'
    }];

    const stats = await runLoader('simple', options, preRules);

    expectToMatchSnapshot(stats);
  });

  it('loads a simple cargo project with wasm-gc', async () => {
    jest.setTimeout(100000);

    const options = {
      release: true,
      gc: true,
    };

    const preRules = [{
      loader: 'wasm-loader'
    }];

    const stats = await runLoader('simple', options, preRules);

    expectToMatchSnapshot(stats);
  });

  it('loads a cargo-web project', async () => {
    jest.setTimeout(100000);

    const options = {
      release: true,
      cargoWeb: true,
      name: 'static/wasm/[name].[hash:8].wasm',
    };

    const stats = await runLoader('stdweb', options);

    expectToMatchSnapshot(stats);
  });
});

function expectToMatchSnapshot(stats) {
  stats = stats.toJson();
  expect(stats.modules.map(m => m.source)).toMatchSnapshot();
}

function runLoader(fixture, options, preRules = []) {
  const config = {
    devtool: 'sourcemap',
    context: path.resolve(__dirname, 'fixtures'),
    entry: `./${fixture}`,
    output: {
      path: path.resolve(__dirname, 'outputs', fixture),
      filename: '[name].bundle.js',
    },
    module: {
      rules: preRules.concat([{
        test: /\.rs$/,
        use: {
          loader: path.resolve(__dirname, '../src'),
          options,
        }
      }])
    },
    plugins: [
      new webpack.optimize.CommonsChunkPlugin(
        {
          name: ['runtime'],
          minChunks: Infinity
        }
      )
    ]
  };

  const compiler = webpack(config);

  compiler.outputFileSystem = new MemoryFS();

  return new Promise((resolve, reject) => compiler.run((err, stats) => {
    if (err) {
      reject(err);
    } else {
      resolve(stats);
    }
  }))
}
