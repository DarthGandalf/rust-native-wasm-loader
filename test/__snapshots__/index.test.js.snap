// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`rust-native-wasm-loader loads a cargo-web project: errors 1`] = `Array []`;

exports[`rust-native-wasm-loader loads a cargo-web project: output 1`] = `"Hello, World!"`;

exports[`rust-native-wasm-loader loads a cargo-web project: warnings 1`] = `Array []`;

exports[`rust-native-wasm-loader loads a simple cargo project with warnings: errors 1`] = `Array []`;

exports[`rust-native-wasm-loader loads a simple cargo project with warnings: output 1`] = `1`;

exports[`rust-native-wasm-loader loads a simple cargo project with warnings: warnings 1`] = `
Array [
  "ModuleWarning: warning: unused variable: \`b\`
 --> src/lib.rs:2:20",
]
`;

exports[`rust-native-wasm-loader loads a simple cargo project with wasm-gc: errors 1`] = `Array []`;

exports[`rust-native-wasm-loader loads a simple cargo project with wasm-gc: output 1`] = `3`;

exports[`rust-native-wasm-loader loads a simple cargo project with wasm-gc: warnings 1`] = `Array []`;

exports[`rust-native-wasm-loader loads a simple cargo project: errors 1`] = `Array []`;

exports[`rust-native-wasm-loader loads a simple cargo project: output 1`] = `3`;

exports[`rust-native-wasm-loader loads a simple cargo project: warnings 1`] = `Array []`;

exports[`rust-native-wasm-loader loads a wasm-bindgen project: errors 1`] = `Array []`;

exports[`rust-native-wasm-loader loads a wasm-bindgen project: output 1`] = `""`;

exports[`rust-native-wasm-loader loads a wasm-bindgen project: warnings 1`] = `Array []`;
